### PR TITLE
samples: add number crunching using optimized library sample

### PR DESCRIPTION
**Pull request by iuliana-prodan**
This sample executes some mathematical functions that can be used for audio processing like
filtering (Fast Fourier Transform (FFT)) or echo cancellation (Least Mean Square (LMS) filter algorithm).

This is an example that demonstrates how to include a proprietary static library into the Zephyr build system.
The library is in an out of tree location.

To use this sample, with an out of tree library, one needs to define an environment variable LIB_LOCATION.
If that is not defined, a default Zephyr API is used instead.

By default it uses CMSIS-DSP backend for math operations.
In README one can find more information.

### CODE REVIEW

**Comment by kartben** [comment]
I can smell a ZDS talk being prepared :)

**Comment by iuliana-prodan** [comment]
> I can smell a ZDS talk being prepared :)

yes, yes... I want this to be an example :)

**Comment by iuliana-prodan** [comment]
@sylvioalves can you, please, help me a bit here?
Why do I get this error on all esp32 platforms:
```
[594/599] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /__w/zephyr/zephyr/twister-out/heltec_wireless_stick_lite_v3_esp32s3_appcpu/samples/application_development/number_crunching/sample.app_dev.number_crunching/zephyr/zephyr_pre0.map 
: && ccache /opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-gcc  -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.obj  -fuse-ld=bfd  -T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/__w/zephyr/zephyr/twister-out/heltec_wireless_stick_lite_v3_esp32s3_appcpu/samples/application_development/number_crunching/sample.app_dev.number_crunching/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/lib/libc/picolibc/liblib__libc__picolibc.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/timer/libdrivers__timer.a  modules/cmsis-dsp/libmodules__cmsis-dsp.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0"  -L/__w/zephyr/zephyr/twister-out/heltec_wireless_stick_lite_v3_esp32s3_appcpu/samples/application_development/number_crunching/sample.app_dev.number_crunching/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -lgcc  -Wl,--fatal-warnings  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,-no-pie  -DPICOLIBC_DOUBLE_PRINTF_SCANF  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/src/linker/esp32s3.rom.alias.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.api.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.libgcc.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.newlib.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.version.ld  -T/__w/zephyr/modules/hal/espressif/zephyr/esp32s3/../../components/soc/esp32s3/ld/esp32s3.peripherals.ld  -Wl,--wrap=longjmp  --specs=picolibc.specs  -lc  -lgcc && cd /__w/zephyr/zephyr/twister-out/heltec_wireless_stick_lite_v3_esp32s3_appcpu/samples/application_development/number_crunching/sample.app_dev.number_crunching/zephyr && /usr/bin/cmake -E true
/opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/zephyr_pre0.elf section `rodata' will not fit in region `dram0_0_seg'
/opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/zephyr_pre0.elf section `rodata' will not fit in region `iram0_0_seg'
/opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: DRAM segment data does not fit.
/opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: region `iram0_0_seg' overflowed by 49112 bytes
/opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: region `dram0_0_seg' overflowed by 90064 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
Should I disable this sample for esp32 or do you have any hint on how to fix this?
Thank you!

**Comment by sylvioalves** [comment]
@iuliana-prodan, Hi, let me check.

**Comment by sylvioalves** [comment]
@iuliana-prodan, I guess there is something that can be done regarding esp32xx appcpu to fix this. However, I think the easiest way right now is for you to add ` arch_exclude: xtensa` from this sample? It seems the code is intended to ARM devices only due to `#include "arm_math.h"`, correct?

**Comment by iuliana-prodan** [comment]
> @iuliana-prodan, I guess there is something that can be done regarding esp32xx appcpu to fix this. However, I think the easiest way right now is for you to add ` arch_exclude: xtensa` from this sample? It seems the code is intended to ARM devices only due to `#include "arm_math.h"`, correct?

This runs also on HiFi4 DSP from i.MX8MP, which is also Xtensa. 
Actually I developed this sample for that target.
By default is using CMSIS-DSP but I also link with NatureDSP from Cadence.
I have an output of the console in the [README](https://github.com/zephyrproject-rtos/zephyr/pull/71079/files#diff-dc1004e1ff8bc2aa2d7499adfd00dd8aa7c52c004ea5c5312032d81fd4dc45ebR55).
So, no, I don't want to disable for Xtensa.

**Comment by sylvioalves** [comment]
> > @iuliana-prodan, I guess there is something that can be done regarding esp32xx appcpu to fix this. However, I think the easiest way right now is for you to add ` arch_exclude: xtensa` from this sample? It seems the code is intended to ARM devices only due to `#include "arm_math.h"`, correct?
> 
> This runs also on HiFi4 DSP from i.MX8MP, which is also Xtensa. Actually I developed this sample for that target. By default is using CMSIS-DSP but I also link with NatureDSP from Cadence. I have an output of the console in the [README](https://github.com/zephyrproject-rtos/zephyr/pull/71079/files#diff-dc1004e1ff8bc2aa2d7499adfd00dd8aa7c52c004ea5c5312032d81fd4dc45ebR55). So, no, I don't want to disable for Xtensa.

Got it.

**Comment by sylvioalves** [comment]
Weird, it SKIPS some test due to RAM overlfow:
` yd_esp32/esp32/appcpu     samples/application_development/number_crunching/sample.app_dev.number_crunching SKIPPED (dram0_1_seg overflow)`
and fails in others with same the reason. Still checking it.

**Comment by iuliana-prodan** [comment]
> @iuliana-prodan This fixes it: #72013 CI is not properly detecting the overflow scenario in there.

Thank you @sylvioalves for the quick fix!

**Comment by dleach02** [comment]
Are we going to try to merge this PR?

**Comment by iuliana-prodan** [comment]
@kartben @nashif please review. Thanks!

**Review by dleach02 — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]
LGTM - although if the vector length were smaller, I would still approve

**Comment by dleach02** [comment]
@kartben need the assignee to review/approve

**Comment by iuliana-prodan** [comment]
@kartben can you please have a look at this?
I want to catch 3.7.

File: samples/application_development/number_crunching/README.rst
Comment (kartben) [inline samples/application_development/number_crunching/README.rst:16]:
```suggestion
an environnent variable ``LIB_LOCATION``.
```

File: samples/application_development/number_crunching/README.rst
Comment (kartben) [inline samples/application_development/number_crunching/README.rst:20]:
```suggestion
In that location, one needs to have a ``CMakeLists.txt``that
defines:

.. code-block:: cmake
```

File: samples/application_development/number_crunching/README.rst
Comment (kartben) [inline samples/application_development/number_crunching/README.rst:40]:
```suggestion
  if :kconfig:option:`CONFIG_CMSIS_DSP` is defined and ``LIB_LOCATION`` is not defined;
```

File: samples/application_development/number_crunching/README.rst
Comment (kartben) [inline samples/application_development/number_crunching/README.rst:43]: this is problematic as this comes with hardly any context. Yes, the point of the sample is probably to showcase proprietary/3rd party integration, but as such the sample will be of little use to folks (myself included) not familiar with what NatureDSP is in the first place. Please see my overrall "review comment" for suggested alternative.

File: samples/application_development/number_crunching/README.rst
Comment (kartben) [inline samples/application_development/number_crunching/README.rst:4]: I suggest using `:zephyr:code-sample:` directive (see guidelines in `doc/templates/sample.tmpl`) -- this might also help actually refine the scope/goal for this sample when completing the "This code sample shows how to ..." sentence

**Review by kartben — CHANGES_REQUESTED** [review]
@iuliana-prodan I am really sorry for taking so long to finally get to this. I think this needs more work as right now the sample is IMO trying to accomplish two very different things, and as such is not doing the best job at either. 

Please let me know if I got it wrong or if you disagree but it feels like trying to both provide CMake sample for how to link to external lib **and** a showcase of "advanced DSP magic with multiple backends, including proprietary ones", is making things unnecessarily complex.

If the sample is meant to be the former, it does indeed belong to "application_development" samples but should really be minimal, and easier to reproduce (i.e. I believe the sample's test should even be able to exercise both options of LIB_LOCATION provided / not provided), and should be named something like "3rd party library integration" or sth like that. But this might be what "external_library" sample does already?

If the sample is meant to be a demonstrator for a generic "number crunching" / "advanced DSP"  I'd say maybe put it under modules/cmsis_dsp as an advanced sample?

**Comment by iuliana-prodan** [comment]
> @iuliana-prodan I am really sorry for taking so long to finally get to this. I think this needs more work as right now the sample is IMO trying to accomplish two very different things, and as such is not doing the best job at either.
> 
> Please let me know if I got it wrong or if you disagree but it feels like trying to both provide CMake sample for how to link to external lib **and** a showcase of "advanced DSP magic with multiple backends, including proprietary ones", is making things unnecessarily complex.
> 
> If the sample is meant to be the former, it does indeed belong to "application_development" samples but should really be minimal, and easier to reproduce (i.e. I believe the sample's test should even be able to exercise both options of LIB_LOCATION provided / not provided), and should be named something like "3rd party library integration" or sth like that. But this might be what "external_library" sample does already?
> 
> If the sample is meant to be a demonstrator for a generic "number crunching" / "advanced DSP" I'd say maybe put it under modules/cmsis_dsp as an advanced sample?

Hi @kartben,
Thanks for the review.
I would like to keep the sample as is, so I will move it in /samples/modules/cmsis_dsp/number_crunching.

File: samples/application_development/number_crunching/README.rst
Comment (iuliana-prodan) [reply samples/application_development/number_crunching/README.rst]: I've moved the sample in samples/modules/cmsis_dsp/number_crunching/.
NatureDSP is the CMSIS-DSP for Xtensa. This can be taken from here: https://github.com/foss-xtensa/ndsplib-hifi4/tree/main, compiled and linked with this sample as I explained in the readme.

**Review by dleach02 — DISMISSED** [review]

**Review by kartben — CHANGES_REQUESTED** [review]
This is a lot of code that's really far from a "minimal" sample. In fact, it feels like it might be designed as a test rather than a sample based on the terminology used in the code? Have you considered making it a test instead?

Also, given we already have the external lib sample and a basic CMSIS-DSP sample, I am not sure this additional sample is needed but I will let the CMSIS-DSP maintainers also chime in.

**Comment by iuliana-prodan** [comment]
> This is a lot of code that's really far from a "minimal" sample. In fact, it feels like it might be designed as a test rather than a sample based on the terminology used in the code? Have you considered making it a test instead?
> 
> Also, given we already have the external lib sample and a basic CMSIS-DSP sample, I am not sure this additional sample is needed but I will let the CMSIS-DSP maintainers also chime in.

Yes, is a complex sample because it not just for ARM. 
The scope is to be used for other architectures also, like Xtensa.

This is an example for NatureDSP from Cadence, for Xtensa arch. 
CMSIS-DSP is used by default since it's available in Zephyr. 
With this sample we can demonstrate how to use equivalent or other APIs that are not in Zephyr.

I don't want to be a test, we already have zdsp tests.
For example, I can't even run the zdsp tests on my DSP from i.MX8MP, which is a HiFi4 Xtensa arch (+ getting errors when trying to compile it with other toolchain - see [here](https://github.com/zephyrproject-rtos/zephyr/issues/70063))

Is a complex sample since, on a DSP, we usually do complex match calculations. 
Also, the API is not always the same for ARM, Xtensa, or others.

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

**Review by andyross — DISMISSED** [review]
Agreed that 1000 lines is pushing the limits of what a "sample" should be, but I guess you could spin that as it being a sample of a non-trivial app and it is after all **interesting** code.

**Review by cfriedt — DISMISSED** [review]
Refresh +1

**Review by dbaluta — DISMISSED** [review]

**Comment by iuliana-prodan** [comment]
@stephanosio @kartben can you, please, have a look at this?
Since Zephyr 3.7 was release, this can be merged now.
Thanks!

**Review by stephanosio — COMMENTED** [review]
This "CMSIS-DSP sample" has its own DSP library abstraction layer and is actually a sample for multiple DSP libraries, including CMSIS-DSP and NatureDSP.

A better (and more correct) way of doing this would be to add the relevant APIs to the Zephyr DSP subsystem modelled after the CMSIS-DSP API, implement the CMSIS-DSP and NatureDSP backends for it, and write a sample for the Zephyr DSP subsystem.

Nevertheless, this still technically is a "CMSIS-DSP sample" and is only a sample; so, I will not leave a -1 if you prefer to not implement the more complete solution described above.

**Comment by iuliana-prodan** [comment]
> This "CMSIS-DSP sample" has its own DSP library abstraction layer and is actually a sample for multiple DSP libraries, including CMSIS-DSP and NatureDSP.
> 
> A better (and more correct) way of doing this would be to add the relevant APIs to the Zephyr DSP subsystem modelled after the CMSIS-DSP API, implement the CMSIS-DSP and NatureDSP backends for it, and write a sample for the Zephyr DSP subsystem.
> 
> Nevertheless, this still technically is a "CMSIS-DSP sample" and is only a sample; so, I will not leave a -1 if you prefer to not implement the more complete solution described above.

Hi @stephanosio, 
I would prefer this to be merged as is.
Your solution can be done later since the backends for CMSIS-DSP and Nature DSP are not quite 1:1 match.
Thanks!

**Comment by kartben** [comment]
I agree with @stephanosio that if the goal is showing generic DSP API it would be better to work towards making this an actual generic front-end in Zephyr DSP subsystem.

I don't think having "advanced" CMSIS-DSP sample is really needed: there's nothing Zephyr specific in the sample as it's "just" showing users how to use CMSIS-DSP API. There's already https://docs.zephyrproject.org/latest/samples/modules/cmsis_dsp/moving_average/README.html showing how to get going with CMSIS-DSP (i.e. how to pull in the module, etc.).

The way forward to make it possible for end users to actually leverage multiple DSP backends is as @stephanosio suggests, not by creating an ad-hoc code sample.

**Comment by iuliana-prodan** [comment]
As I previously mentioned [here ](https://github.com/zephyrproject-rtos/zephyr/pull/71079#issuecomment-2172387585)this is not just a sample for CMSIS-DSP.

This is an example to demonstrate that one can start developing a sample from the existing code in Zephyr (like CMSIS-DSP) and then port it or integrate it with a 3rt party library or a proprietary code. I know we have the application_development/external_lib, but that doesn’t do much, is just a dummy code.

If I were to add a sample with 2 backends I would keep only simple operations (like vector addition) since the other math ops are no exact match between CMSIS-DSP and NatureDSP.
I also tried ZDSP tests – see my comment [here](https://github.com/zephyrproject-rtos/zephyr/pull/71079#issuecomment-2172387585).
The frontend of ZDSP does not match NatureDSP, only a few vector operations. 
So, to make a generic frontend is a complicated task, that cannot be addressed by this PR.

With this sample I want to demonstrate that you don’t always have a 1:1 match, you have to make some tweaks in order to have your code integrated with an existing sample.

Since, you don’t consider this a well defined application I will move it in samples/boards/nxp_adsp.

**Comment by iuliana-prodan** [comment]
Changes since last version:
- move the sample from `samples/modules/cmsis_dsp/number_crunching/` to `samples/boards/nxp/adsp/number_crunching/`.

**Comment by dleach02** [comment]
@kartben we would like to get this sample merged into the upstream. What is needed to address your change request?

**Review by kartben — CHANGES_REQUESTED** [review]

File: samples/boards/nxp/adsp/number_crunching/sample.yaml
Comment (kartben) [inline samples/boards/nxp/adsp/number_crunching/sample.yaml:8]: this is missing a `platform_allow:` entry -- we don't want to build samples on all platforms as this just kills CI, and only those boards that the sample have actually been tested on and that are advertised in the README should be there
https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html#definition-and-criteria

**Review by kartben — COMMENTED** [review]
sorry, not sure why my second comment did not make it

**Review by kartben — COMMENTED** [review]
sorry, not sure why my second comment did not make it

File: samples/boards/nxp/adsp/number_crunching/sample.yaml
Comment (iuliana-prodan) [reply samples/boards/nxp/adsp/number_crunching/sample.yaml:11]: done

File: samples/boards/nxp/adsp/number_crunching/README.rst
Comment (iuliana-prodan) [reply samples/boards/nxp/adsp/number_crunching/README.rst]: Hope is better this way: `This sample shows how to set up and use different backends for complex math operations.` :)

File: samples/boards/nxp/adsp/number_crunching/README.rst
Comment (kartben) [inline samples/boards/nxp/adsp/number_crunching/README.rst:4]:
```suggestion
   Set up and use different backends for complex math operations.
```

**Review by kartben — CHANGES_REQUESTED** [review]

**Review by kartben — COMMENTED** [review]
+1 for docs

**Review by kartben — APPROVED** [review]
+1 for docs

**Comment by iuliana-prodan** [comment]
> Changes since last version:
> 
> * move the sample from `samples/modules/cmsis_dsp/number_crunching/` to `samples/boards/nxp/adsp/number_crunching/`.

@stephanosio can you, please, have a look at this. 
As seen above I've moved the sample in `samples/boards/nxp/adsp/number_crunching/`.
Thanks!

**Review by stephanosio — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]

**Pull request merged**
